### ISSUE TITLE
scylla crashes when select * with secondary index on large dataset

### ISSUE THREAD

**Issue by jwnx** [scylladb/scylla#4147]
**Installation details**
Scylla version: 3.0.1
Cluster size: 3
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0e408f1b170700dd0

Scylla has a lot of reactor stalls until it finally crashes. 
(Possibly related to #3726)

**Steps to reproduce**:
1. Yaml:
```
keyspace: indexed 
keyspace_definition: |
  CREATE KEYSPACE indexed WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1};

table: buildings
table_definition: |
  CREATE TABLE buildings (
    name text,
    city text,
    built int,
    feet int,
    PRIMARY KEY (name))
    WITH compression = {}

columnspec:
  - name: name
    cluster: fixed(1000)
    population: uniform(0..1000M)

  - name: city
    size: fixed(1000)
    population: gaussian(0..1000)

  - name: built 
    size: fixed(1000)
    population: gaussian(1940..2019, 1990, 20)

  - name: feet 
    size: fixed(1000)
    population: gaussian(30..2800, 200, 50)

insert:
  partitions: fixed(1)  
  batchtype: UNLOGGED            
  select: fixed(2000)/2000    

queries:
  read1:
    cql: select * from users where username = ? and startdate = ?
    fields: samerow
```

2. Insert 1B keys (now with nohup, because this will take time)
```
nohup cassandra-stress user profile=buildings.yaml n=1000000000 ops\(insert=1\) no-warmup -node NODE_IP -rate 'threads=450' -mode 'native cql3 connectionsPerHost=30' >/dev/null 2>&1 &
```

3. Query `cqlsh --request-timeout 3600 -e "paging off; select * from indexed.buildings where feet=203 and built=1962 ALLOW FILTERING;"`. It will timeout after 1 hour. No warnings.

4. Create a secondary index: `create index on indexed.buildings (feet);`

5. Query `cqlsh --request-timeout 3600 -e "paging off; select * from indexed.buildings where feet=203 and built=1962 ALLOW FILTERING;"`. It will return the following and eventually crash a node.
```
 :1:ReadTimeout: Error from server: code=1200 [Coordinator node timed out waiting for replica nodes' responses] message="Operation timed out for indexed.buildings - received only 0 responses from 1 CL=ONE." info={'received_responses': 0, 'required_responses': 1, 'consistency': 'ONE'}
```

Logs attached. Index creation starts at 10:01:13 and finishes at 11:50:17. Querying begin at 13:06:00.

[lots-of-data-crash.log](https://github.com/scylladb/scylla/files/2796504/lots-of-data-crash.log)

**Comment by duarten** [comment]
Are those the logs of the server that crashed? I don't see anything in the logs, but I guess it ran out of memory. It would be nice to have a backtrace or a core dump. It would also be interesting to decode some of those stalls and oversized allocations.

Did you try with paging on? We should behave better (although there's still #3808). Nobody should be doing unpaged queries, and it's very sad the protocol allows them. The coordinator needs to materialize the whole result set in memory, so it may not be possible to answer those queries. Still, we shouldn't crash. In the [design document](https://docs.google.com/document/d/1MZa3Y6BiMaqbIj43QAo2q0U7rZJtWJdnDbfHlC_sJDQ) I suggested we page unpaged queries internally, which can allow us to detect whether we should at some point abort the query, either because we're running out of memory or are over the client-specified timeout (related: #3809).

Fortunately, range queries are also not the typical use-case for secondary indexes.

**Comment by glommer** [comment]
On Fri, Jan 25, 2019 at 12:53 PM Duarte Nunes <notifications@github.com>
wrote:

> Are those the logs of the server that crashed? I don't see anything in the
> logs, but I guess it ran out of memory. It would be nice to have a
> backtrace or a core dump. It would also be interesting to decode some of
> those stalls and oversized allocations.
>
> Did you try with paging on? We should behave better (although there's
> still #3808 <https://github.com/scylladb/scylla/issues/3808>). Nobody
> should be doing unpaged queries, and it's very sad the protocol allows
> them. The coordinator needs to materialize the whole result set in memory,
> so it may not be possible to answer those queries. Still, we shouldn't
> crash. In the design document
> <https://docs.google.com/document/d/1MZa3Y6BiMaqbIj43QAo2q0U7rZJtWJdnDbfHlC_sJDQ>
> I suggested we page unpaged queries internally, which can allow us to
> detect whether we should at some point abort the query, either because
> we're running out of memory or are over the client-specified timeout
> (related: #3809 <https://github.com/scylladb/scylla/issues/3809>).
>
> Fortunately, range queries are also not the typical use-case for secondary
> indexes.\\
>
they are with filtering

**Comment by duarten** [comment]
Why is that? Filtered range queries on secondary indexes should be even rarer (and sometimes better done as filtering queries on the base table directly: #3810).

**Comment by slivne** [comment]
The timeout of the first query 
```
cqlsh --request-timeout 3600 -e "paging off; select * from indexed.buildings where feet=203 and built=1962 ALLOW FILTERING;"
```

timeout was from the client side (1h) - why didn't it timeout on the server side ? if it hasn't timed out on the server side it could keep on running even after the client has timed out - which would mean the attempt with SI was done in parallel to an ongoing query eating up memory ion the background. 

AFAIK clients can not request a different request timeout - so its server side configured, as such if the default was used in scylla the first query should have timed out long before the 1h.

**Comment by kostja** [comment]
@avikivity I am inclined to close this as not a bug. If you fetch a large data set, don't turn off the paging. Alternatively, we could short-cut execution without paging when the result set size exceeds K pages, K being a configuration constant.
wdyt?
